src/pal/inc/pal.h

```diff
@@ -5971,129 +5941,82 @@ class PAL_NoHolder
 public:
     void SuppressRelease() {}
 };
+
+struct PAL_SEHException
+{
+public:
+    // Note that the following two are actually embedded in this heap-allocated
+    // instance - in contrast to Win32, where the exception record would usually
+    // be allocated on the stack.  This is needed because foreign cleanup handlers
+    // partially unwind the stack on the second pass.
+    EXCEPTION_POINTERS ExceptionPointers;
+    EXCEPTION_RECORD ExceptionRecord;
+    CONTEXT ContextRecord;
+
+    PAL_SEHException(EXCEPTION_RECORD *pExceptionRecord, CONTEXT *pContextRecord)
+    {
+        ExceptionPointers.ExceptionRecord = &ExceptionRecord;
+        ExceptionPointers.ContextRecord = &ContextRecord;
+        ExceptionRecord = *pExceptionRecord;
+        ContextRecord = *pContextRecord;
+    }
+};
+
 #endif // __cplusplus
 
 #define PAL_TRY(__ParamType, __paramDef, __paramRef)                            \
 {                                                                               \
-    struct __HandlerData                                                        \
-    {                                                                           \
-        __ParamType __param;                                                    \
-        EXCEPTION_DISPOSITION __handlerDisposition;                             \
-        __HandlerData(__ParamType param) : __param(param) {}                    \
+    __ParamType __param = __paramRef;                                           \
+    auto tryBlock = [](__ParamType __paramDef)                                  \
+    {
+
+#define PAL_EXCEPT(dispositionExpression)                                       \
     };                                                                          \
-    __HandlerData __handlerData(__paramRef);                                    \
-    class __Body                                                                \
+    const bool isFinally = false;                                                     \
+    auto finallyBlock = []() {};                                                \
+    try                                                                         \
+    {                                                                           \
+        tryBlock(__param);                                                      \
+    }                                                                           \
+    catch (PAL_SEHException& ex)                                                \
```

Comment:
I think we should catch(...) in addition to catching (PAL_SEHException) - to match the Windows behavior. Of course, we are not going to have CONTEXT* for the non-SEHException. NULL CONTEXT* should be ok for most cases, if not all.